### PR TITLE
#80 Add xml import feature implementation

### PR DESCRIPTION
**Pull request by LionnelC**
If you tell me how to test, I can provide some.
Are you using a containered version of Polarion  ?

### CODE REVIEW

**Comment by jesper-raemaekers** [comment]
I do have a docker image containing Polarion with a configuration that works for the the tests in this project. I cannot share it due to the license restrictions that come with Polarion, but if you are familiar with Linux I could give you access to the server where it is installed.

File: docs/xml.rst [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+Import an xml file of results
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Import an xml file of results (see xml_junit.xsd for the standard format) into a new or an existing test run in a project
+
+
+.. code:: python
+
+    import os
+    from polarion.xml import Config, Importer
```
Comment (jesper-raemaekers) [inline docs/xml.rst:19]: Add ResultExporter import here

File: docs/xml.rst [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        logging.basicConfig(level=logging.INFO)
+        testrun_comment='comment to set in the test run' 
+
+        config=Config.from_dict({
+            Config.XML_FILE: 'xml_file.xml',
+            Config.URL: 'http://hostname/polarion',
+            Config.TOKEN: 'secret_token', # can also use USERNAME and PASSWORD
+            Config.PROJECT_ID: 'project_id_in_polarion',
+            Config.TESTRUN_ID: 'testrun_id_in_polarion', # if not set, create a new test run
+            Config.TESTRUN_COMMENT: 'comment to add in the test run' # as an option.
```
Comment (jesper-raemaekers) [inline docs/xml.rst:32]: Also add SKIP_MISSING_TESTCASE here.

File: polarion/xml.py [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+    TESTRUN_ID_GENERATOR='testrun_id_generator'         # function that generate the testrun_id (config as parameter) if testrun_id is not provided. If not set: f'unit-{datetime.now}'
+    TESTRUN_TITLE='testrun_title'                       # title of the test run if testrun is created. If not set 'New unit test run'
+    TESTRUN_TYPE='testrun_type'                         # type of the test run if testrun is created. If not set 'xUnit Test Manual Upload'
+    TESTRUN_COMMENT='testrun_comment'                   # test run comment to add if set.
+    SKIP_MISSING_TESTCASE='skip_missing_testcase'       # if set to True, skip result on unknown test cases
+    ATTRIBUTES = [
+        XML_FILE, URL, USERNAME, PASSWORD, TOKEN, 
+        PROJECT_ID, TESTRUN_ID, TESTRUN_ID_GENERATOR,
+        TESTRUN_TITLE, TESTRUN_TYPE, TESTRUN_COMMENT]
+    MANDATORY = [XML_FILE, PROJECT_ID]  # and also either user/password or token
```
Comment (jesper-raemaekers) [inline polarion/xml.py:28]: SKIP_MISSING_TESTCASE is missing here. It would seem there need to be more MANDATORY items.

File: docs/xml.rst [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+            })
+
+        testrun=Importer.from_xml(config)
+        if config.result is not None:
+            ResultExporter.save_json(config.result, testrun)
```
Comment (jesper-raemaekers) [inline docs/xml.rst:37]: this bit is not working for me as config.result does not exists

**Review by jesper-raemaekers — CHANGES_REQUESTED** [review]

**Comment by LionnelC** [comment]
> I do have a docker image containing Polarion with a configuration that works for the the tests in this project. I cannot share it due to the license restrictions that come with Polarion, but if you are familiar with Linux I could give you access to the server where it is installed.

Jesper, I understand the license restrictions. I may have the same on my side ;)
Do you think it is possible to share the dockerfile to build it ?

File: polarion/xml.py (hunk 3 above)
Comment (LionnelC) [reply polarion/xml.py:28]: Ok for SKIP_MISSING_TESTCASE missing as ATTRIBUTES.
I add URL as mandatory. Token or user/password are manually check as put in comment.
I do not see any other mandatory to check:
- TESTRUN_ID is optional. If not set it created a new one at each call
- TESTRUN_ID_GENERATOR has a default value
- TESTRUN_TITLE has a default value
- TESTRUN_TYPE has a default value
- TESTRUN_COMMENT is optional
- SKIP_MISSING_TESTCASE is optional

File: docs/xml.rst (hunk 1 above)
Comment (LionnelC) [reply docs/xml.rst]: @jesper-raemaekers, anything is missing ?

**Review by jesper-raemaekers — APPROVED** [review]

**Comment by LionnelC** [comment]
Great !

**Comment by jesper-raemaekers** [comment]
I think sharing the docker file should be fine. there are some other minor modifications that are also needed. Probably it is best to discuss this via Discord or something.

I have noticed a new docker that does appear to start polarion: https://hub.docker.com/r/boaz1/polarion
I have not tested it and i don't know if it is allowed to share that container.

**Comment by jesper-raemaekers** [comment]
@LionnelC shall i make a release from it now or do you want to create some tests first?

**Comment by LionnelC** [comment]
> I think sharing the docker file should be fine. there are some other minor modifications that are also needed. Probably it is best to discuss this via Discord or something.
> 
> I have noticed a new docker that does appear to start polarion: https://hub.docker.com/r/boaz1/polarion I have not tested it and i don't know if it is allowed to share that container.

Here is my discord; lionnel0 #1945

**Comment by LionnelC** [comment]
> @LionnelC shall i make a release from it now or do you want to create some tests first?

You can release it. I've performed many manual tests & internal demo. I may write some tests later

**Pull request merged**